Fixes #26338 - use port from HTTP boot feature

lzap opened the pull request:
When PXE loader with "HTTP BOOT" is used, Foreman need to set DHCP boot
option to an URL instead of a filename. Smart proxy deployed with HTTP
Boot feature exposes TFTP directory via HTTP and HTTPS however it uses
nonstandard ports (8443/8000). Those ports cannot be detected from
Foreman DB, until now. Smart proxy now have a capabilities and settings
exposing feature:

https://github.com/theforeman/smart-proxy/pull/676

This allows Foreman to figure out ports which can be then used in DHCP
orchestration to set the correct DHCP URL.

lzap committed:
Fixes #26338 - use port from HTTP boot feature

ekohl commented on app/models/operatingsystem.rb:
```diff
@@ -255,8 +255,14 @@ def boot_filename(host = nil)
-    boot_uri = URI.parse((host.subnet&.httpboot?) ? host.subnet.httpboot.url : Setting[:unattended_url])
-    self.class.all_loaders_map(architecture, "#{boot_uri.host}:#{boot_uri.port}")[host.pxe_loader]
+    if host.subnet&.httpboot?
+      port = host.subnet.httpboot.setting(:HTTPBoot, 'http_port') || 80
```
Why port 80? Do you assume Foreman is running on the same host or did you intend `8000`?

It is possible that `http_port` is set but the module is https-only. That means you should also check the `http_enabled` setting. Perhaps we should amend the smart-proxy side to ensure the port is nil unless http_enabled is true?

lzap replied:
> Why port 80?

My thinking is that user wants to run her own HTTP server, which is usually 80. Another option is to raise an error and insist on setting being exposed which probably makes more sense and creates less confusion.

> It is possible that http_port is set but the module is https-only.

That is not possible, I tested this. In that case `http_port` is `nil`.

lzap replied:
On the other hand, defaulting to `8000` enables use with older smart proxies.

ekohl replied:
> > It is possible that http_port is set but the module is https-only.
> 
> That is not possible, I tested this. In that case `http_port` is `nil`

It is. Reversed, but I tested this in the proxy by writing an integration test:

```
def test_features_http
  Proxy::DefaultModuleLoader.any_instance.expects(:load_configuration_file).with('httpboot.yml').returns(enabled: 'http', root_dir: '/var/lib/tftpboot')

  get '/features'

  response = JSON.parse(last_response.body)

  mod = response['httpboot']
  refute_nil(mod)
  assert(mod['http_enabled'])
  refute(mod['https_enabled'])
  assert_equal('running', mod['state'], Proxy::LogBuffer::Buffer.instance.info[:failed_modules][:httpboot])
  assert_equal([], mod['capabilities'])

  expected_settings = {'http_port' => nil, 'https_port' => 8443}
  assert_equal(expected_settings, mod['settings'])
end
```

Note `https_port` is set, but `https_enabled` is false because the module is configured with`enabled: 'http'`. That was what I was thinking. IMHO we should fix this on the proxy side so the `*_port` values are nil when it's not enabled so you don't have to think about this on the Foreman side.

> On the other hand, defaulting to `8000` enables use with older smart proxies.

I think this is more logical. We don't really support custom HTTP servers.

lzap replied:
I was confused by this:

```ruby
    def http_enabled?
      !settings.http_port.nil?
    end
```

I thought it's the same setting and haven't considered that whole httpboot module can be enabled via `enabled: 'http'`. I agree, let me change this on the proxy side then.

ekohl replied:
Currently we can't distinguish between a proxy that doesn't have the settings exposed (too old) and doesn't have it enabled. For that we need to check if the key is set. I'm slightly leaning to not maintaining compatibility with older proxies since HTTP boot is a relative new feature. Then we show a clear error to the user. Do you agree we should drop the defaults on the Foreman side?

lzap replied:
https://github.com/theforeman/smart-proxy/pull/677

lzap commented:
Fixed tests, small change in ports actually - I realized that I must set HTTP or HTTPS according to the PXE loader setting. Added tests around this as well. Also incorporated your comment - it defaults to our default ports just in case user is still on unupgraded smart proxy version.

ekohl commented on app/models/operatingsystem.rb:
```diff
@@ -255,8 +255,18 @@ def boot_filename(host = nil)
-    self.class.all_loaders_map(architecture, "#{boot_uri.host}:#{boot_uri.port}")[host.pxe_loader]
+    if host.subnet&.httpboot?
+      if host.pxe_loader =~ /UEFI HTTPS/
+        port = host.subnet.httpboot.setting(:HTTPBoot, 'https_port') || 8443
```
Since it differs for Katello I'm wondering if we should try to detect this based on the `host.subnet.httpboot.url`. If its scheme is https then we can use that port as a default. On the other hand, that's rather complex and I'd also be fine with hard requiring a proxy that exposes this setting which would mean a hard exception + a note in the upgrade notes. Slightly leaning to that and avoid the guessing.

lzap replied:
What makes you believe that user uses HTTPS for smart proxy API communication? For example in my development setup I use HTTP so I can track the communication using wireshark.

ekohl replied:
I'd rather hard error if we can't come up with a sane default than try to guess. As an admin that's easier to debug. So essentially `host.subnet.httpboot.setting(:HTTPBoot, 'https_port') || raise`

lzap replied:
Ok I can do that.

lzap commented:
Force-pushed.

lzap commented:
Fixed tests.

This is the last patch to get UEFI HTTP Booting: 
https://community.theforeman.org/t/uefi-http-boot-via-grub2/15255

ekohl commented on app/models/operatingsystem.rb:
```diff
@@ -255,8 +255,18 @@ def boot_filename(host = nil)
-    self.class.all_loaders_map(architecture, "#{boot_uri.host}:#{boot_uri.port}")[host.pxe_loader]
+    if host.subnet&.httpboot?
+      if host.pxe_loader =~ /UEFI HTTPS/
+        port = host.subnet.httpboot.setting(:HTTPBoot, 'https_port') || raise(::Foreman::Exception.new(N_("HTTPS boot requires proxy with httpboot feature and https_port exposed setting")))
```
One question I have is about `:HTTPBoot` vs a string. Does it matter?

timogoebel replied:
I'm wondering if it would be better to invalidate the host object (via a rails validation) in addition to throwing an exception here. The exception is not very user-friendly imho.

lzap replied:
> One question I have is about `:HTTPBoot` vs a string. Does it matter?

I am not following.

> invalidate the host object

Good idea, lemme see if it's easy change.

lzap replied:
Hmmm the code is used only in orchestration and there we tend to use exceptions as the orchestration stack tries to catch them and roll back. So I'd probably keep it as-is.

lzap replied:
> One question I have is about :HTTPBoot vs a string. Does it matter?

If you mean that I should use 'HTTPBoot' instead, I can change it if you insist but ActiveRecord apparently works fine with that.

tbrisker replied:
What's the outcome of this discussion? is this good to merge or do you plan to change it?

lzap replied:
The ball is on @ekohl side, thanks!

lzap replied:
What is the status @ekohl ? If there is a change you want me to do let me know and I will happily do it. This is a customer-wanted feature, thanks!

lzap replied:
Again, I can change the `:HTTPBoot` to string if that's what @ekohl wants, I am not sure. Please confirm.

ekohl reviewed (approved):
Overall :+1: and I think it can be merged. Not a blocker to merge, but I'm wondering if we also have exception codes in Foreman.

I think this also could use an upgrade note that when using the HTTPBoot feature, the proxy must be 1.24+. This is ok with me given how little the feature is used since it's so new.

lzap commented:
During testing today I have found that grub2 can have problem with explicit port in URL (e.g. `xxx:8000`) and the only option until this is fixed is to forward 80->8000 on the proxy. Would you mind a setting that would override the port returned from proxy actually? @ekohl 

Also please explain what you mean in the other thread and I will hapilly rebase, I just don't understand.

ares commented:
After quick assurance of the current state, I'm also OK merging as is. People simply need 1.24 on both sides in order to have this working, so trying to support old proxy doesn't make much sense. It's good it already uses capabilities for feature detection.

Thanks @lzap, merging!

frippe75 commented:
I'm on Foreman 1.24.1 and still getting "Oops, we're sorry but something went wrong ERF42-9666 [Foreman::Exception]: HTTP boot requires proxy with httpboot feature and http_port exposed setting"

I'm simply trying to setup a PXE boot scenaio using a Smart Proxy and when I try to create the host I get this errror message. I cannot see a POST for tftp service on the smart proxy node.
